File: python/xrprimer/data_structure/camera/camera.py
import copy
import logging
import os
from typing import Union

import numpy as np

from xrprimer.utils.log_utils import get_logger
from xrprimer_cpp.camera import BaseCameraParameter as BaseCameraParameter_cpp


class BaseCameraParameter(BaseCameraParameter_cpp):
    ATTR_NAMES = [
        'name', 'intrinsic', 'extrinsic_r', 'extrinsic_t', 'height', 'width',
        'world2cam', 'convention'
    ]

    def __init__(self,
                 K: Union[list, np.ndarray, None] = None,
                 R: Union[list, np.ndarray, None] = None,
                 T: Union[list, np.ndarray, None] = None,
                 name: str = 'default',
                 height: int = 1080,
                 width: int = 1920,
                 world2cam: bool = True,
                 convention: str = 'opencv',
                 logger: Union[None, str, logging.Logger] = None) -> None:
        """Base class for all python camera parameter classes. Common methods
        are defined in this class.

        Args:
            K (Union[list, np.ndarray, None], optional):
                Nested list of float32, 4x4 or 3x3 K mat.
                Defaults to None, 4x4 zeros.
            R (Union[list, np.ndarray, None], optional):
                Nested list of float32, 3x3 rotation mat.
                Defaults to None, 3x3 identity.
            T (Union[list, np.ndarray, None], optional):
                List of float32, T vector.
                Defaults to None, zero vector.
            name (str, optional):
                Name of this camera. Defaults to 'default'.
            height (int, optional):
                Height of the image shot by this camera.
                Defaults to 1080.
            width (int, optional):
                Width of the image shot by this camera.
                Defaults to 1920.
            world2cam (bool, optional):
                Whether the R, T transform points from world space
                to camera space. Defaults to True.
            convention (str, optional):
                Convention name of this camera.
                Defaults to 'opencv'.
            logger (Union[None, str, logging.Logger], optional):
                Logger for logging. If None, root logger will be selected.
                Defaults to None.
        """
        super().__init__(
            intrinsic=np.zeros((4, 4)),
            extrinsic_r=np.zeros((3, 3)),
            extrinsic_t=np.ones((3, )))
        self.name = name
        self.set_KRT(K, R, T)
        self.set_resolution(height=height, width=width)
        self.world2cam = world2cam
        self.convention = convention
        self.logger = get_logger(logger)

    def set_intrinsic(self,
                      mat3x3: Union[list, np.ndarray, None] = None,
                      width: int = None,
                      height: int = None,
                      fx: float = None,
                      fy: float = None,
                      cx: float = None,
                      cy: float = None,
                      perspective: bool = True) -> None:
        """Set the intrinsic of a camera. Note that mat3x3 has a higher
        priority than fx, fy, cx, cy.

        Args:
            mat3x3 (list, optional):
                A nested list of intrinsic matrix,
                in shape (3, 3). If mat is given,
                fx, fy, cx, cy will be ignored.
                Defaults to None.
            width (int):
                Width of the screen.
            height (int):
                Height of the screen.
            fx (float, optional):
                Focal length. Defaults to None.
            fy (float, optional):
                Focal length. Defaults to None.
            cx (float, optional):
                Camera principal point. Defaults to None.
            cy (float, optional):
                Camera principal point. Defaults to None.
            perspective (bool, optional):
                Whether it is a perspective camera, if not,
                it's orthographics. Defaults to True.
        """
        if mat3x3 is not None:
            mat3x3 = np.asarray(mat3x3)
            super().set_intrinsic(mat3x3=mat3x3, perspective=perspective)
        elif width is not None and\
                height is not None and\
                fx is not None and\
                fy is not None and\
                cx is not None and\
                cy is not None:
            super().set_intrinsic(
                width=width, height=height, fx=fx, fy=fy, cx=cx, cy=cy)
        else:
            self.logger.error(
                'Either mat3x3 or (h, w, fx/y, cx/y) should be offered.')
            raise ValueError

    def set_resolution(self, height: int, width: int) -> None:
        """Set resolution of the camera.

        Args:
            height (int):
                Height of the screen.
            width (int):
                Width of the screen.
        """
        self.height = height
        self.width = width

    def set_KRT(self,
                K: Union[list, np.ndarray, None] = None,
                R: Union[list, np.ndarray, None] = None,
                T: Union[list, np.ndarray, None] = None,
                world2cam: Union[bool, None] = None) -> None:
        """Set K, R to matrix and T to vector.

        Args:
            K (Union[list, np.ndarray, None]):
                Nested list of float32, 4x4 or 3x3 K mat.
                Defaults to None, intrisic will not be changed.
            R (Union[list, np.ndarray, None]):
                Nested list of float32, 3x3 R mat.
                Defaults to None, extrisic_r will not be changed.
            T (Union[list, np.ndarray, None]):
                List of float32, T vector.
                Defaults to None, extrisic_t will not be changed.
            world2cam (Union[bool, None], optional):
                Whether the R, T transform points from world space
                to camera space.
                Defaults to None, self.world2cam will not be changed.
        """
        if K is not None:
            if len(K) == 4:
                self.intrinsic = np.asarray(K)
            else:
                self.set_intrinsic(mat3x3=K, perspective=True)
        if R is not None:
            self.extrinsic_r = np.asarray(R)
        if T is not None:
            self.extrinsic_t = np.asarray(T)
        if world2cam is not None:
            assert isinstance(world2cam, bool)
            self.world2cam = world2cam

    def inverse_extrinsic(self) -> None:
        """Inverse the direction of extrinsics, between world to camera and
        camera to world."""
        r_mat = np.asarray(self.extrinsic_r)
        t_vec = np.asarray(self.extrinsic_t)
        r_mat = np.linalg.inv(r_mat).reshape(3, 3)
        t_vec = -np.dot(r_mat, t_vec).reshape(3)
        self.extrinsic_r = r_mat
        self.extrinsic_t = t_vec
        self.world2cam = not self.world2cam

    def intrinsic33(self) -> np.ndarray:
        """Get an intrinsic matrix in shape (3, 3).

        Returns:
            ndarray: An ndarray of intrinsic matrix.
        """
        return super().intrinsic33()

    def get_intrinsic(self, k_dim: int = 3) -> list:
        """Get intrinsic K matrix.

        Args:
            k_dim (int, optional):
                If 3, returns a 3x3 mat.
                Else if 4, returns a 4x4 mat.
                Defaults to 3.

        Raises:
            ValueError: k_dim is neither 3 nor 4.

        Returns:
            list: Nested list of float32, 4x4 or 3x3 K mat.
        """
        if k_dim == 4:
            return self.intrinsic.tolist()
        elif k_dim == 3:
            return super().intrinsic33().tolist()
        else:
            self.logger.error('k_dim is neither 3 nor 4.')
            raise ValueError

    def get_extrinsic_r(self) -> list:
        """Get extrinsic rotation matrix.

        Returns:
            list: Nested list of float32, 3x3 R mat.
        """
        return self.extrinsic_r.tolist()

    def get_extrinsic_t(self) -> list:
        """Get extrinsic translation vector.

        Returns:
            list: Nested list of float32, T vec of length 3.
        """
        return self.extrinsic_t.reshape(3).tolist()

    def SaveFile(self, filename: str) -> int:
        """Dump camera name and parameters to a json file.

        Args:
            filename (str):
                Path to the dumped json file.

        Returns:
            int: returns 0.
        """
        return BaseCameraParameter_cpp.SaveFile(self, filename)

    def dump(self, filename: str) -> None:
        """Dump camera name and parameters to a json file.

        Args:
            filename (str):
                Path to the dumped json file.

        Raises:
            RuntimeError: Fail to dump a json file.
        """
        if not self.SaveFile(filename):
            self.logger.error('Fail to dump a json file.')
            raise RuntimeError

    def LoadFile(self, filename: str) -> bool:
        """Load camera name and parameters from a dumped json file.

        Args:
            filename (str):
                Path to the dumped json file.

        Returns:
            bool: True if load succeed.
        """
        return BaseCameraParameter_cpp.LoadFile(self, filename)

    def load(self, filename: str) -> None:
        """Load camera name and parameters from a dumped json file.

        Args:
            filename (str):
                Path to the dumped json file.

        Raises:
            FileNotFoundError: File not found at filename.
            ValueError: Content in filename is not correct.
        """
        if not os.path.exists(filename):
            self.logger.error(f'File not found at {filename}.')
            raise FileNotFoundError
        test_cam = self.__class__()
        load_test = test_cam.LoadFile(filename)
        if load_test:
            self.LoadFile(filename)
        else:
            self.logger.error('File content is not correct.')
            raise ValueError

    def clone(self) -> 'BaseCameraParameter':
        """Clone a new CameraPrameter instance like self.

        Returns:
            BaseCameraParameter
        """
        new_cam_param = self.__class__(
            K=copy.deepcopy(self.get_intrinsic(k_dim=4)),
            R=copy.deepcopy(self.extrinsic_r),
            T=copy.deepcopy(self.extrinsic_t),
            name=self.name,
            height=self.height,
            width=self.width,
            world2cam=self.world2cam,
            convention=self.convention,
            logger=self.logger)
        return new_cam_param

    @classmethod
    def fromfile(cls, filename: str) -> 'BaseCameraParameter':
        """Construct a camera parameter data structure from a json file.

        Args:
            filename (str):
                Path to the dumped json file.

        Returns:
            CameraParameter:
                An instance of CameraParameter class.
        """
        ret_cam = cls()
        ret_cam.load(filename)
        return ret_cam

File: cpp/pybind/data_structure/camera.cpp

#include "pybind/data_structure/camera.h"
#include <data_structure/camera/camera.h>
#include <data_structure/camera/fisheye_camera.h>
#include <data_structure/camera/omni_camera.h>
#include <data_structure/camera/pinhole_camera.h>

// Base
class PyBaseCameraParameter : public BaseCameraParameter {
public:
  using BaseCameraParameter::BaseCameraParameter;

  std::string ClassName() const override {
    PYBIND11_OVERRIDE_PURE(std::string, BaseCameraParameter, ClassName, );
  }

  bool SaveFile(const std::string &filename) const override {
    PYBIND11_OVERRIDE_PURE(bool, BaseCameraParameter, SaveFile, filename);
  }

  bool LoadFile(const std::string &filename) override {
    PYBIND11_OVERRIDE_PURE(bool, BaseCameraParameter, LoadFile, filename);
  }
};

// Pinhole
class PyPinholeCameraParameter : public PinholeCameraParameter {
public:
  using PinholeCameraParameter::PinholeCameraParameter;

  std::string ClassName() const override {
    PYBIND11_OVERRIDE(std::string, PinholeCameraParameter, ClassName, );
  }

  bool SaveFile(const std::string &filename) const override {
    PYBIND11_OVERRIDE(bool, PinholeCameraParameter, SaveFile, filename);
  }

  bool LoadFile(const std::string &filename) override {
    PYBIND11_OVERRIDE(bool, PinholeCameraParameter, LoadFile, filename);
  }
};

// Omni
class PyOmniCameraParameter : public OmniCameraParameter {
public:
  using OmniCameraParameter::OmniCameraParameter;

  std::string ClassName() const override {
    PYBIND11_OVERRIDE(std::string, OmniCameraParameter, ClassName, );
  }

  bool SaveFile(const std::string &filename) const override {
    PYBIND11_OVERRIDE(bool, OmniCameraParameter, SaveFile, filename);
  }

  bool LoadFile(const std::string &filename) override {
    PYBIND11_OVERRIDE(bool, OmniCameraParameter, LoadFile, filename);
  }
};

// Fisheye
class PyFisheyeCameraParameter : public FisheyeCameraParameter {
public:
  using FisheyeCameraParameter::FisheyeCameraParameter;

  std::string ClassName() const override {
    PYBIND11_OVERRIDE(std::string, FisheyeCameraParameter, ClassName, );
  }

  bool SaveFile(const std::string &filename) const override {
    PYBIND11_OVERRIDE(bool, FisheyeCameraParameter, SaveFile, filename);
  }

  bool LoadFile(const std::string &filename) override {
    PYBIND11_OVERRIDE(bool, FisheyeCameraParameter, LoadFile, filename);
  }
};

// pybind modules
void pybind_camera_classes(py::module &m) {
  // Base
  py::class_<BaseCameraParameter, PyBaseCameraParameter>(m,
                                                         "BaseCameraParameter")
      .def(py::init<const Eigen::Matrix4f &, const Eigen::Matrix3f &,
                    const Eigen::Vector3f &>(),
           "intrinsic"_a, "extrinsic_r"_a, "extrinsic_t"_a)
      .def("ClassName", &BaseCameraParameter::ClassName)
      .def("SaveFile", &BaseCameraParameter::SaveFile)
      .def("LoadFile", &BaseCameraParameter::LoadFile)
      .def("set_intrinsic",
           static_cast<void (BaseCameraParameter::*)(int, int, double, double,
                                                     double, double, bool)>(
               &BaseCameraParameter::set_intrinsic),
           "width"_a, "height"_a, "fx"_a, "fy"_a, "cx"_a, "cy"_a,
           "perspective"_a = true)
      .def("set_intrinsic",
           static_cast<void (BaseCameraParameter::*)(const Eigen::Matrix3f &,
                                                     bool)>(
               &BaseCameraParameter::set_intrinsic),
           "mat3x3"_a, "perspective"_a = true)
      .def("intrinsic33", &BaseCameraParameter::intrinsic33)
      .def_property(
          "intrinsic",
          [](BaseCameraParameter &p) -> Eigen::Ref<Eigen::Matrix4f> {
            return p.intrinsic_;
          },
          [](BaseCameraParameter &p, const Eigen::Matrix4f &mat) {
            p.intrinsic_ = mat;
          },
          "camera intrinsic (4x4): numpy.ndarray[numpy.float32[4, 4]] or list")
      .def_property(
          "extrinsic_r",
          [](BaseCameraParameter &p) -> Eigen::Ref<Eigen::Matrix3f> {
            return p.extrinsic_r_;
          },
          [](BaseCameraParameter &p, const Eigen::Matrix3f &mat) {
            p.extrinsic_r_ = mat;
          },
          "camera extrinsics R: numpy.ndarray[numpy.float32[3, 3]] or list")
      .def_property(
          "extrinsic_t",
          [](BaseCameraParameter &p) -> Eigen::Ref<Eigen::Vector3f> {
            return p.extrinsic_t_;
          },
          [](BaseCameraParameter &p, const Eigen::Vector3f &vec) {
            p.extrinsic_t_ = vec;
          },
          "camera extrinsics T: numpy.ndarray[numpy.float32[3, 1]] or list")
      .def_readwrite("name", &BaseCameraParameter::name_, "camera tag name")
      .def_readwrite("height", &BaseCameraParameter::height_,
                     "camera image height: int")
      .def_readwrite("width", &BaseCameraParameter::width_,
                     "camera image width: int")
      .def_readwrite("convention", &BaseCameraParameter::convention_,
                     "transform convention, default is opencv: str")
      .def_readwrite("world2cam", &BaseCameraParameter::world2cam_,
                     "world to camera flag: bool");

  // Pinhole
  py::class_<PinholeCameraParameter, BaseCameraParameter,
             PyPinholeCameraParameter>(m, "PinholeCameraParameter")
      .def(py::init<>(), "PinholeCameraParameter constructor");

  // Omni
  py::class_<OmniCameraParameter, BaseCameraParameter, PyOmniCameraParameter>(
      m, "OmniCameraParameter")
      .def(py::init<>(), "OmniCameraParameter constructor")
      .def_readwrite("k1", &OmniCameraParameter::k1_, " : float ")
      .def_readwrite("k2", &OmniCameraParameter::k2_, " : float ")
      .def_readwrite("k3", &OmniCameraParameter::k3_, " : float ")
      .def_readwrite("k4", &OmniCameraParameter::k4_, " : float ")
      .def_readwrite("k5", &OmniCameraParameter::k5_, " : float ")
      .def_readwrite("k6", &OmniCameraParameter::k6_, " : float ")
      .def_readwrite("p1", &OmniCameraParameter::p1_, " : float ")
      .def_readwrite("p2", &OmniCameraParameter::p2_, " : float ")
      .def_readwrite("xi", &OmniCameraParameter::xi_, " : float ")
      .def_property(
          "D",
          [](OmniCameraParameter &p) -> Eigen::Ref<Eigen::Vector4f> {
            return p.D_;
          },
          [](OmniCameraParameter &p, const Eigen::Vector4f &vec) {
            p.D_ = vec;
          },

          "omni camera D: numpy.ndarray[numpy.float32[4, 1]] or list");

  // Fisheye
  py::class_<FisheyeCameraParameter, BaseCameraParameter,
             PyFisheyeCameraParameter>(m, "FisheyeCameraParameter")
      .def(py::init<>(), "FisheyeCameraParameter constructor")
      .def_readwrite("k1", &FisheyeCameraParameter::k1_, " : float")
      .def_readwrite("k2", &FisheyeCameraParameter::k2_, " : float")
      .def_readwrite("k3", &FisheyeCameraParameter::k3_, " : float")
      .def_readwrite("k4", &FisheyeCameraParameter::k4_, " : float")
      .def_readwrite("k5", &FisheyeCameraParameter::k5_, " : float")
      .def_readwrite("k6", &FisheyeCameraParameter::k6_, " : float")
      .def_readwrite("p1", &FisheyeCameraParameter::p1_, " : float")
      .def_readwrite("p2", &FisheyeCameraParameter::p2_, " : float");
}

void xrprimer_pybind_camera(py::module &m) {
  py::module m_submodule = m.def_submodule("camera");
  pybind_camera_classes(m_submodule);
}
